### ISSUE TITLE
Missing assignement during `rows.Err()` invocation

### ISSUE THREAD

**Issue by silktrader** [sapienzaapps/wasa-fountains#2]
There may be an issue with the following statement:

```go
	if rows.Err() != nil {
		return nil, err
	}
```

https://github.com/sapienzaapps/wasa-fountains/blob/a0d8a778cf2a1150a36d9c1325afec79246154c1/service/database/list-fountains-with-filter.go#L61

I may be mistaken but, assuming the execution got to that point, the variable `err` will likely be ```nil```, meaning that the latter will be returned, in lieu of ```rows.Err()``` return value. An alternative:

```go
	if err = rows.Err(); err != nil {
		return nil, err
	}
```

**Comment by silktrader** [comment]
_Not part of the report_

This isn't the place, but I seize the occasion to seek feedback ...

I often instantiate an empty return slice, as with ```var fountains = make([]Fountain, 0)```, which I return during every error check, rather than ```nil```. The idea is to have partial results, in case of failure (as could happen inside the for loop). The initial assignment cost is negligible. Are there dangers or better alternatives?

**Comment by silktrader** [comment]
_Not part of the report_

The author is checking whether fountain points fall within a given circle. I assume one could offload the haversine distance calculations to the DB and not process extra rows. In this case, though, with SQLite eager to block the DB at every turn, wouldn't it be (needlessly) more efficient to perform slice filtering outside of the rows iteration, to release the lock faster?

**Comment by Enrico204** [comment]
> There may be an issue with the following statement:
> 
> ```go
> 	if rows.Err() != nil {
> 		return nil, err
> 	}
> ```
> 
> I may be mistaken but, assuming the execution got to that point, the variable `err` will likely be `nil`, meaning that the latter will be returned, in lieu of `rows.Err()` return value. An alternative:
> 
> ```go
> 	if err = rows.Err(); err != nil {
> 		return nil, err
> 	}
> ```

You're correct. The second statement was the intended one; I'll correct it ASAP.

> _Not part of the report_
> 
> This isn't the place, but I seize the occasion to seek feedback ...
> 
> I often instantiate an empty return slice, as with `var fountains = make([]Fountain, 0)`, which I return during every error check, rather than `nil`. The idea is to have partial results, in case of failure (as could happen inside the for loop). The initial assignment cost is negligible. Are there dangers or better alternatives?

In Go, the best practice is to check the error before using every other return value, and if it's not `nil` (i.e., an error occurred), other parameters should be ignored as their content is undefined. Of course, this can be "amended", meaning that the function/package documentation may specify otherwise.

If there is an error, I avoid returning partial results to stick to the standard library behavior. However, there are no dangers in returning partial values AFAIK, except for the partial result itself (which may not be a problem, depending on the context of the code).

> _Not part of the report_
> 
> The author is checking whether fountain points fall within a given circle. I assume one could offload the haversine distance calculations to the DB and not process extra rows. In this case, though, with SQLite eager to block the DB at every turn, wouldn't it be (needlessly) more efficient to perform slice filtering outside of the rows iteration, to release the lock faster?

I suppose that you can find a way of tweaking SQLite to calculate the haversine distance directly. However, to answer also to the second part of your comment, keep in mind that:
* SQLite is working in memory, so, in practice, the math is being processed by the same machine (no advantages like offloading the math to the SQL server)
* For the same reason as before, no waste of network bandwidth (except maybe some memory allocation and disk I/O)
* The only case when this may be a problem is when a huge number of fountains are between the square and the circle (i.e., near the vertexes, inside the square), which is rare for an example project
* Moving the math in SQL may be a form of [premature optimization](https://softwareengineering.stackexchange.com/a/80092)
* Also, we don't have SQL in the WASA syllabus, so this is another point toward keeping most of the logic in Go

The moment a project reaches a critical point (hopefully, before that point, thanks to metrics) in terms of performance, solutions like the one you're proposing may be implemented. In fact, there are already defined functions or extensions for SQL servers with optimized Geospatial functions, like [Postgis](https://postgis.net/).

**Comment by Enrico204** [comment]
Fixed in commit 27d61646fa9d099adfe1991fb957821ef61c87b0. If you have any further question, please ask them during lectures (let's use issues for bug reporting, so it's easy for us to handle everything), or via e-mail if you can't attend.
